### PR TITLE
Response header hooks

### PR DESCRIPTION
**Pull request by apparentlymart**
The API for this client library is focused on returning the primary data associated with each API response, which typically means just the body of the response.

Sometimes clients will also need to react to cross-cutting metadata such as expiration times, cache control guidance, and rate limiting information, which isn't a direct part of the data being requested but can nonetheless affect the behavior of the client. This information is typically returned in HTTP response header fields.

To give access to this information without a breaking change to the API, this uses the `context.Context` API to allow the rare caller that needs it to register a hook through which it will be notified about the response header in any case where the request succeeded enough for there to be one.

Most clients will not need this facility, which justifies the light abuse of the `context.Context` API for passing in this optional hook, even though this isn't the sort of cross-cutting concern `context.Context` should typically be used for.

---

The goal of introducing this now is to improve hashicorp/terraform#32983 so that the Terraform Cloud / Enterprise API can return a new response header signaling its preferred state snapshot interval, thereby creating a compromise between creating no intermediate snapshots at all vs. creating so many intermediate snapshots as to overwhelm the storage of Terraform Enterprise customers and run across potential race conditions in Terraform Cloud.

This is a compromise to have a sort of relief valve in case of problems. This does not necessarily replace other efforts to store and present intermediate snapshots in a better way in Terraform Cloud, although it could potentially allow working on those with less urgency.

### CODE REVIEW

**Comment by apparentlymart** [comment]
I can see that the "errcheck" lint has failed, but the failure doesn't seem to have generated any information about the source location of the problem. :thinking:

It looks like the job is intended to generate annotations for the source code to mark where the problems are -- I see it asking the tool to generate its results in a GitHub Actions-specific formation -- but no annotations have appeared. I'm not sure what exactly that lint failure is referring to.

File: request_hooks.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+func ContextWithResponseHeaderHook(parentCtx context.Context, cb func(status int, header http.Header)) context.Context {
+	// If the given context already has a notification callback then we'll
+	// arrange to notify both the previous and the new one. This is not
+	// a super efficient way to achieve that but we expect it to be rare
+	// for there to be more than one or two hooks associated with a particular
+	// request, so it's not warranted to optimize this further.
+	existingI := parentCtx.Value(contextResponseHeaderHookKey)
+	finalCb := cb
+	if existingI != nil {
+		existing := existingI.(func(int, http.Header))
```
Comment (brandonc) [inline request_hooks.go:35]: This was the failing errcheck. I'm not sure why this isn't showing up in CI, so I'll check into that.

```
$ make lint
golangci-lint run .
request_hooks.go:35:15: Error return value is not checked (errcheck)
		existing := existingI.(func(int, http.Header))
```

**Review by brandonc — DISMISSED** [review]
This is a nice solution! A test infrastructure dependency (run tasks mock) is down right now so I will merge this manually once the linter passes

File: request_hooks.go (hunk 1 above)
Comment (apparentlymart) [reply request_hooks.go:35]: :thinking: interesting...

So is the linter rule here that it's just totally forbidden to use the one-result form of type assertion?

It looks like `golangci-lint` is wrapping [`errcheck`](https://github.com/kisielk/errcheck/blob/8f86be0cd56a95f4b0ddcbab3eec22f04e58b202/errcheck/errcheck.go#L620-L635) here and indeed it does seem to have [a rather-onerous rule that single-result type assertions are just never allowed](https://github.com/kisielk/errcheck/blob/8f86be0cd56a95f4b0ddcbab3eec22f04e58b202/errcheck/errcheck.go#L620-L635).

The only way this assertion could fail is if another part of this file were buggy, so I'm going to try to resolve this by just _explicitly_ panicking if the assertion fails, thereby achieving exactly the same effect as what I originally wrote but hopefully in a way that will satisfy the linter.

**Review by brandonc — APPROVED** [review]

**Pull request merged**